### PR TITLE
STOR-1290: don't pass credentials via environment variables in vSphere

### PR DESCRIPTION
**Pull request by bertinatto**
CC @openshift/storage

### CODE REVIEW

File: pkg/operator/vspherecontroller/driver_starter.go [hunk 1]
```diff
@@ -113,7 +107,6 @@ func (c *VSphereController) createCSIDriver() {
 			c.apiClients.ConfigMapInformer.Informer(),
 			c.apiClients.NodeInformer.Informer(),
 		},
-		WithVSphereCredentials(defaultNamespace, cloudCredSecretName, c.infraLister, c.configMapLister, c.apiClients.SecretInformer),
```
Comment (bertinatto) [inline pkg/operator/vspherecontroller/driver_starter.go:116]: Credentials are not longer set as env variables in the container, so I don't need a function hook anymore.

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 2]
```diff
@@ -717,3 +724,64 @@ func (c *VSphereController) createStorageClassController() storageclasscontrolle
...
+	usernameKey = vCenterName + ".username"
+	passwordKey = vCenterName + ".password"
+
+	if usernameKey == "" || passwordKey == "" {
+		return "", "", fmt.Errorf("could not find vSphere credentials in secret %s/%s", secret.Namespace, secret.Name)
+	}
+
+	// Get user and pass from CCO secret
+	return string(secret.Data[usernameKey]), string(secret.Data[passwordKey]), nil
+}
```
Comment (bertinatto) [inline pkg/operator/vspherecontroller/vspherecontroller.go:762]: I converted the hook into this function and moved it to this file.

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 3]
```diff
@@ -717,3 +724,64 @@ func (c *VSphereController) createStorageClassController() storageclasscontrolle
...
+	if !ok {
+		return "", fmt.Errorf("cloud config %s/%s does not contain key %q", cloudConfigNamespace, cloudConfig.Name, cloudConfig.Key)
+	}
+	cfg := new(vsphere.VSphereConfig)
+	err = gcfg.ReadStringInto(cfg, cfgString)
+	if err != nil {
+		return "", err
+	}
+	return cfg.Workspace.VCenterIP, nil
+}
```
Comment (bertinatto) [inline pkg/operator/vspherecontroller/vspherecontroller.go:787]: This function has just moved from `driver_starter.go`to here.

**Comment by bertinatto** [comment]
Since this is changing the way the driver authenticates in vSphere, I want to make sure this doesn't break vSphere jobs:

/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-csi
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-serial
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.17-upgrade-from-stable-4.16-e2e-vsphere-ovn-upgrade
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-upi
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-upi-serial
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-sdn
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-static-ovn

**Comment by bertinatto** [comment]
/test e2e-vsphere

`[sig-etcd] etcd should not log excessive took too long messages`

File: assets/csi_cloud_config.ini [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 [Global]
 cluster-id = "${CLUSTER_ID}"
+user = "${USER}"
+password = "${PASSWORD}"
```
Comment (jsafrane) [inline assets/csi_cloud_config.ini:6]: Do we need to escape weird characters in the password? I assume env. var could have anything except '\0' and we did not need to care about escaping there. What are rules in the ini file?

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 5]
```diff
@@ -633,13 +637,13 @@ func (c *VSphereController) createCSIConfigMap(
...
 
-	requiredCM, err := c.applyClusterCSIDriverChange(infra, cfg, clusterCSIDriver, datastoreURL)
+	requiredSecret, err := c.applyClusterCSIDriverChange(infra, cfg, clusterCSIDriver, datastoreURL)
 	if err != nil {
 		return err
 	}
 
 	// TODO: check if configMap has been deployed and set appropriate conditions
-	_, _, err = resourceapply.ApplyConfigMap(ctx, c.kubeClient.CoreV1(), syncContext.Recorder(), requiredCM)
+	_, _, err = resourceapply.ApplySecret(ctx, c.kubeClient.CoreV1(), syncContext.Recorder(), requiredSecret)
```
Comment (gnufied) [inline pkg/operator/vspherecontroller/vspherecontroller.go:646]: What happens in upgraded clusters? Do we need to remove the previously configured configmap?

File: pkg/operator/vspherecontroller/driver_starter.go [hunk 6]
```diff
@@ -302,12 +210,12 @@ func WithLogLevelDaemonSetHook() csidrivernodeservicecontroller.DaemonSetHookFun
...
 
-func WithConfigMapDaemonSetAnnotationHook(configMapName, namespace string, configMapInformer corev1informers.ConfigMapInformer) csidrivernodeservicecontroller.DaemonSetHookFunc {
+func WithSecretDaemonSetAnnotationHook(secretName, namespace string, secretInformer corev1informers.SecretInformer) csidrivernodeservicecontroller.DaemonSetHookFunc {
 	return func(opSpec *operatorapi.OperatorSpec, ds *appsv1.DaemonSet) error {
 		inputHashes, err := resourcehash.MultipleObjectHashStringMapForObjectReferenceFromLister(
-			configMapInformer.Lister(),
 			nil,
-			resourcehash.NewObjectRef().ForConfigMap().InNamespace(namespace).Named(configMapName),
+			secretInformer.Lister(),
+			resourcehash.NewObjectRef().ForSecret().InNamespace(namespace).Named(secretName),
```
Comment (gnufied) [inline pkg/operator/vspherecontroller/driver_starter.go:218]: Strictly speaking, do we even need this anymore? We are not mounting the csi-config in daemonset anyways...

File: assets/csi_cloud_config.ini (hunk 4 above)
Comment (bertinatto) [reply assets/csi_cloud_config.ini:6]: Escaping backslashes and quotation marks.

File: pkg/operator/vspherecontroller/vspherecontroller.go (hunk 5 above)
Comment (bertinatto) [reply pkg/operator/vspherecontroller/vspherecontroller.go:652]: Added logic to remove legacy ConfigMap

File: pkg/operator/vspherecontroller/driver_starter.go (hunk 6 above)
Comment (bertinatto) [reply pkg/operator/vspherecontroller/driver_starter.go:218]: I'm not sure; do you remember why we added this? The only ref I could find was this: https://github.com/openshift/vmware-vsphere-csi-driver-operator/pull/107/files#diff-1d5594c6562bc9ce4771e4821f10c8580165298d4067db516eb283385676e5efR122

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 7]
```diff
@@ -717,3 +733,95 @@ func (c *VSphereController) createStorageClassController() storageclasscontrolle
...
+	}
+
+	// Get username and pass from secret created by CCO
+	username := string(secret.Data[usernameKey])
+	password := string(secret.Data[passwordKey])
+
+	// The CSI driver expects a password with any quotation marks and backslashes escaped.
+	// xref: https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/121
+	password = reEscape.ReplaceAllString(password, `\$0`)
+	return username, escapeQuotesAndBackslashes(password), nil
```
Comment (gnufied) [inline pkg/operator/vspherecontroller/vspherecontroller.go:776]: Are we escaping this two times? didn't we just escape this above?

File: pkg/operator/vspherecontroller/driver_starter.go (hunk 6 above)
Comment (bertinatto) [reply pkg/operator/vspherecontroller/driver_starter.go:220]: Added a comment, PTAL

File: pkg/operator/vspherecontroller/vspherecontroller.go [hunk 8]
```diff
@@ -633,13 +643,13 @@ func (c *VSphereController) createCSIConfigMap(
 
 	datastoreURL := defaultDatastore.Summary.Url
 
-	requiredCM, err := c.applyClusterCSIDriverChange(infra, cfg, clusterCSIDriver, datastoreURL)
```
[the remark this answers is not in the document]
Comment (bertinatto) [reply pkg/operator/vspherecontroller/vspherecontroller.go:636]: Done

File: pkg/operator/vspherecontroller/vspherecontroller.go (hunk 7 above)
Comment (bertinatto) [reply pkg/operator/vspherecontroller/vspherecontroller.go:775]: Good catch; this mistake is a result of a code refactoring where I moved the logic to `escapeQuotesAndBackslashes()` in order to unit-test it. PTAL

File: pkg/operator/vspherecontroller/vspherecontroller.go (hunk 2 above)
Comment (bertinatto) [reply pkg/operator/vspherecontroller/vspherecontroller.go:781]: This comment originally referred to `getUserAndPassword`, not `escapeQuotesAndBackslashes`.

**Comment by gnufied** [comment]
The failing upgrade tests look unrelated.

/lgtm

**Comment by bertinatto** [comment]
/label docs-approved
/label px-approved
This is not a user-facing change.

**Comment by bertinatto** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-csi
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-serial
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-techpreview-serial
/payload-job periodic-ci-openshift-release-master-ci-4.17-upgrade-from-stable-4.16-e2e-vsphere-ovn-upgrade
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-upi
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-ovn-upi-serial
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-sdn
/payload-job periodic-ci-openshift-release-master-nightly-4.17-e2e-vsphere-static-ovn

**Comment by duanwei33** [comment]
I created one vSphere cluster with the open PR, there is no vSphere credentials as environment variables in vmware-vsphere-csi-driver-controller pod, but I noticed the `e2e-vsphere-zones ` CI failed because the CSI Driver pod is not running, see [log](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_vmware-vsphere-csi-driver-operator/227/pull-ci-openshift-vmware-vsphere-csi-driver-operator-master-e2e-vsphere-zones/1808127679883382784/artifacts/e2e-vsphere-zones/gather-extra/artifacts/pods/openshift-cluster-csi-drivers_vmware-vsphere-csi-driver-controller-cf6c85b58-lg4cn_csi-driver.log).
I'm not sure how relative it is but let's have a retest.

**Pull request merged**